mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithm.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.james.mailbox.model.ThreadId;
+import org.apache.james.mailbox.postgres.mail.dao.PostgresThreadDAO;
+import org.apache.james.mailbox.store.mail.ThreadIdGuessingAlgorithm;
+import org.apache.james.mailbox.store.mail.model.MimeMessageId;
+import org.apache.james.mailbox.store.mail.model.Subject;
+import org.apache.james.mailbox.store.search.SearchUtil;
+
+import com.google.common.hash.Hashing;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class PostgresThreadIdGuessingAlgorithm implements ThreadIdGuessingAlgorithm {
+    private final MailboxManager mailboxManager;
+    private final PostgresMailboxFactory mailboxFactory;
+
+    @Inject
+    public PostgresThreadIdGuessingAlgorithm(MailboxManager mailboxManager, PostgresMailboxFactory mailboxFactory) {
+        this.mailboxManager = mailboxManager;
+        this.mailboxFactory = mailboxFactory;
+    }
+
+    @Override
+    public Mono<ThreadId> guessThreadIdReactive(MessageId messageId, Optional<MimeMessageId> mimeMessageId, Optional<MimeMessageId> inReplyTo, Optional<List<MimeMessageId>> references, Optional<Subject> subject, MailboxSession session) {
+        PostgresThreadDAO threadDAO = mailboxFactory.createThreadDAO(session.getUser().getDomainPart());
+
+        Set<Integer> hashMimeMessageIds = buildMimeMessageIdSet(mimeMessageId, inReplyTo, references)
+            .stream()
+            .map(mimeMessageId1 -> Hashing.murmur3_32_fixed().hashBytes(mimeMessageId1.getValue().getBytes()).asInt())
+            .collect(Collectors.toSet());
+
+        Optional<Integer> hashBaseSubject = subject.map(value -> new Subject(SearchUtil.getBaseSubject(value.getValue())))
+            .map(subject1 -> Hashing.murmur3_32_fixed().hashBytes(subject1.getValue().getBytes()).asInt());
+
+        return threadDAO.findThreads(session.getUser(), hashMimeMessageIds)
+            .filter(pair -> pair.getLeft().equals(hashBaseSubject))
+            .next()
+            .map(Pair::getRight)
+            .switchIfEmpty(Mono.just(ThreadId.fromBaseMessageId(messageId)))
+            .flatMap(threadId -> threadDAO
+                .insertSome(session.getUser(), hashMimeMessageIds, messageId, threadId, hashBaseSubject)
+                .then(Mono.just(threadId)));
+    }
+
+    @Override
+    public Flux<MessageId> getMessageIdsInThread(ThreadId threadId, MailboxSession session) {
+        SearchQuery searchQuery = SearchQuery.builder()
+            .andCriteria(SearchQuery.threadId(threadId))
+            .sorts(new SearchQuery.Sort(SearchQuery.Sort.SortClause.Arrival, SearchQuery.Sort.Order.NATURAL))
+            .build();
+
+        MultimailboxesSearchQuery expression = MultimailboxesSearchQuery
+            .from(searchQuery)
+            .build();
+
+        return Flux.from(mailboxManager.search(expression, session, Integer.MAX_VALUE))
+            .switchIfEmpty(Mono.error(() -> new ThreadNotFoundException(threadId)));
```

Comment:
I would love to be able to do this against PG only.
Here I understand this would be done against OpenSearch.

I bet this would be as easy as reading the `thread` table by `username` and `threadId` ?

Reply:
In the current implementation, data will be sorted by arrival date. In postgres implementation, sorting data by MESSAGE_ID could be good option. And to do that, MESSAGE_ID must be generated by UuidCreator.getTimeOrderedEpoch(). WDYT?

Reply:
updated